File: GSP_test/GSPtest/GSP.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace GSPtest
{
  class GSP
  {
    List<Sequence> freq_seqs = new List<Sequence>();
    List<Sequence> cur_freq_seqs = new List<Sequence>();
    Dictionary<int, string> item_map = new Dictionary<int, string>();
    List<Client> clients = new List<Client>();
    double support = 1;
    int min_gap = 0;
    int max_gap = int.MaxValue;
    int win_size = 0;
    private bool IsFrequentSubSeq(Sequence seq)
    {
      List<Sequence> sub_seqs = seq.GetSubSequences();
      List<bool> flags = new List<bool>(sub_seqs.Count);
      for (int i = 0; i < sub_seqs.Count; i++)
        flags.Add(false);
      for (int i = 0; i < sub_seqs.Count; i++)
        foreach (Sequence freq_seq in cur_freq_seqs)
          if (sub_seqs[i] == freq_seq)
            flags[i] = true;
      bool res = true;
      for (int i = 0; i < flags.Count; i++)
        res = res && flags[i];
      return res;
    }
    private List<Sequence> GenerateCandidates1()
    {
      List<Sequence> candidates = new List<Sequence>(0);
      for (int i = 0; i < this.item_map.Count; i++)
        candidates.Add(new Sequence(i + 1));
      return candidates;
    }
    private List<Sequence> GenerateCandidates()
    {
      List<Sequence> candidates = new List<Sequence>(0);
      //Join
      foreach (Sequence seq1 in cur_freq_seqs)
        foreach (Sequence seq2 in cur_freq_seqs)
          if (Sequence.IsJoinable(seq1, seq2))
            if (seq1 != seq2) // exclude <(x),(x)> sequences
              candidates.Add(Sequence.Join(seq1, seq2)); // <(x),(y)>
      if (cur_freq_seqs[0].Size() == 1) //if prev cur_freq_seqs has 1 elem only
        for (int i = 0; i < cur_freq_seqs.Count; i++)
          for (int j = i + 1; j < cur_freq_seqs.Count; j++)
            if (cur_freq_seqs[i].Size() == 1 && cur_freq_seqs[j].Size() == 1)
              candidates.Add(Sequence.Join1(cur_freq_seqs[i], cur_freq_seqs[j])); // <(x, y)>
      //Prune
      for (int i = 0; i < candidates.Count; i++)
      {
        if (!IsFrequentSubSeq(candidates[i]))
          candidates.Remove(candidates[i--]);
      }
      return candidates;
    }
    private bool FindElement(ItemGroup element, Client client, int time, out int start_time, out int end_time)
    {
      LinkedListNode<int>[] nodes_list = client.GetRepresintationNodesList();

      start_time = int.MaxValue;
      end_time = 0;

      for (int i = 0; i < element.Size(); i++)
      {
        int item_id = element[i] - 1;
        if (nodes_list[item_id] == null)
          return false;
        while (nodes_list[item_id].Value < time)
        {
          if (nodes_list[item_id].Next == null)
            return false;
          nodes_list[item_id] = nodes_list[item_id].Next;
        }
        if (nodes_list[item_id].Value < start_time)
          start_time = nodes_list[item_id].Value;
        if (nodes_list[item_id].Value > end_time)
          end_time = nodes_list[item_id].Value;
      }

      //checking win_size
      while (end_time - start_time > win_size)
      {
        int new_start_time = start_time;
        for (int i = 0; i < element.Size(); i++)
        {
          int item_id = element[i] - 1;

          if (nodes_list[item_id].Value == start_time)
          {
            if (nodes_list[item_id].Next == null)
              return false;
            nodes_list[item_id] = nodes_list[item_id].Next;

            if (nodes_list[item_id].Value > start_time)
              new_start_time = nodes_list[item_id].Value;
            /*if (nodes_list[item_id].Value > end_time)
              end_time = nodes_list[item_id].Value;*/
          }
        }

        start_time = new_start_time;

        if (start_time > end_time)
          return false;
      }

      return true;
    }
    private bool ClientSupportsSequence(Client client, Sequence seq)
    {
      bool res = false;
      int phase = 1;
      int cur_id = 0;
      int[] start_times = new int[seq.Size()];
      int[] end_times = new int[seq.Size()];

      while (phase != 0)
      {
        if (cur_id == seq.Size())
        {
          res = true;
          break;
        }
        if (phase == 1) //Forward phase
        {
          bool is_founded = FindElement(seq[cur_id], client, cur_id == 0 ? 0 : end_times[cur_id - 1] + min_gap, out int new_start_time, out int new_end_time);
          if (is_founded)
          {
            start_times[cur_id] = new_start_time;
            end_times[cur_id] = new_end_time;
          }
          if (!is_founded) // no element in transactions
          {
            res = false;
            phase = 0;
            break;
          }
          else if (cur_id == 0 || new_end_time - start_times[cur_id - 1] < max_gap) //ok
          {
            cur_id++;
            phase = 1;
          }
          else // switch to backward phase
          {
            phase = 2;
          }
        }
        else if (phase == 2) //Backward phase
        {
          if (cur_id == 0)
            phase = 1;
          else
          {
            bool is_founded = FindElement(seq[cur_id - 1], client, end_times[cur_id] - max_gap, out int new_start_time, out int new_end_time);
            if (is_founded)
            {
              start_times[cur_id - 1] = new_start_time;
              end_times[cur_id - 1] = new_end_time;
            }
            if (!is_founded) // no element in transactions
            {
              res = false;
              phase = 0;
              break;
            }
            else if (cur_id - 1 == 0 || new_end_time - start_times[cur_id - 2] < max_gap) //ok
            {
              if (end_times[cur_id] - new_start_time > min_gap)
                cur_id++;
              phase = 1;
            }
            else // continue backward phase
            {
              cur_id--;
              phase = 2;
            }
          }
        }
        else //eror phase
        {
          break;
        }
      }

      return res;
    }
    private int CountSupport(List<Sequence> candidates)
    {
      int res = 0;
      cur_freq_seqs = new List<Sequence>();
      foreach (Sequence candidate in candidates)
      {
        int supported_clients = 0;
        foreach (Client client in clients)
        {
          if (ClientSupportsSequence(client, candidate))
            supported_clients++;
        }
        double candidate_support = (double)supported_clients / clients.Count;
        if (candidate_support >= this.support)
        {
          candidate.support = candidate_support;
          candidate.confidence = CalcConfidence(candidate);
          candidate.lift = CalcLift(candidate);
          freq_seqs.Add(candidate);
          cur_freq_seqs.Add(candidate);
          res++;
        }
      }
      return res;
    }
    private Sequence FindFreqSequenceByItemGroup(ItemGroup gr)
    {
      foreach (Sequence seq in freq_seqs)
        if (seq.Size() == 1 && seq[0] == gr)
          return seq;
      return new Sequence();
    }
    private double CalcConfidence(Sequence seq)
    {
      double res = seq.support;
      ItemGroup gr = seq[0];
      Sequence subseq = FindFreqSequenceByItemGroup(gr);
      if (subseq.Size() < 1)
        return -1;
      res /= subseq.support;
      return res;
    }
    private double CalcLift(Sequence seq)
    {
      double res = seq.support;
      foreach (ItemGroup gr in seq)
      {
        Sequence subseq = FindFreqSequenceByItemGroup(gr);
        if (subseq.Size() < 1)
          return -1;
        res /= subseq.support;
      }
      return res;
    }
    public List<Sequence> GetFrequentSequences()
    {
      this.freq_seqs = new List<Sequence>();
      List<Sequence> candidates = GenerateCandidates1();
      int added_seqs_num = CountSupport(candidates);
      if (freq_seqs.Count > 0)
      {
        while (candidates.Count > 0 && added_seqs_num > 0)
        {
          candidates = GenerateCandidates();
          added_seqs_num = CountSupport(candidates);
        }
      }
      SortFrequentSequences();
      return this.freq_seqs;
    }
    public void SortFrequentSequences()
    {
      this.freq_seqs.Sort((x, y) => -x.support.CompareTo(y.support));
      this.freq_seqs.Sort((x, y) => -x.confidence.CompareTo(y.confidence));
      this.freq_seqs.Sort((x, y) => -x.lift.CompareTo(y.lift));
    }
    public int PrintFrequentSequences()
    {
      int max_len = 0;
      for (int i = 0; i < this.freq_seqs.Count; i++)
      {
        string res = "<";
        foreach (ItemGroup gr in this.freq_seqs[i])
        {
          res += "(";
          foreach (int item in gr)
            res += this.item_map[item] + ",";
          res = res.Remove(res.Length - 1, 1);
          res += "),";
        }
        res = res.Remove(res.Length - 1, 1);
        res += ">";
        if (res.Length > max_len)
          max_len = res.Length;

        res += " supp = " + String.Format("{0:F3}", this.freq_seqs[i].support);
        res += " conf = " + (this.freq_seqs[i].confidence != -1 ? String.Format("{0:F3}", this.freq_seqs[i].confidence) : "null ");
        res += " lift = " + (this.freq_seqs[i].lift != -1 ? String.Format("{0:F3}", this.freq_seqs[i].lift) : "null ");

        Console.WriteLine(res);
      }
      return max_len;
    }
    public void WriteFrequentSequencesToFile(string file_name = "gsp_result.txt", int max_len = -1)
    {
      using (StreamWriter file = new StreamWriter(file_name))
      {
        for (int i = 0; i < this.freq_seqs.Count; i++)
        {
          string res = "<";
          foreach (ItemGroup gr in this.freq_seqs[i])
          {
            res += "(";
            foreach (int item in gr)
              res += this.item_map[item] + ",";
            res = res.Remove(res.Length - 1, 1);
            res += "),";
          }
          res = res.Remove(res.Length - 1, 1);
          res += ">";

          while (res.Length < max_len)
            res += " ";

          res += " supp = " + String.Format("{0:F3}", this.freq_seqs[i].support);
          res += " conf = " + (this.freq_seqs[i].confidence != -1 ? String.Format("{0:F3}", this.freq_seqs[i].confidence) : "null ");
          res += " lift = " + (this.freq_seqs[i].lift != -1 ? String.Format("{0:F3}", this.freq_seqs[i].lift) : "null ");

          file.WriteLine(res);
        }
      }
    }
    public void Test1()
    {
      cur_freq_seqs = new List<Sequence>(0);
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 1, 2 } ),
                                                           new ItemGroup( new List<int> { 3 } ) }));
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 1, 2 } ),
                                                           new ItemGroup( new List<int> { 4 } ) }));
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 1 } ),
                                                           new ItemGroup( new List<int> { 3, 4 } ) }));
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 1, 3 } ),
                                                           new ItemGroup( new List<int> { 5 } ) }));
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 2 } ),
                                                           new ItemGroup( new List<int> { 3, 4 } ) }));
      cur_freq_seqs.Add(new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 2 } ),
                                                           new ItemGroup( new List<int> { 3 } ),
                                                           new ItemGroup( new List<int> { 5 } )}));

      List<Sequence> candidates2 = this.GenerateCandidates();
    }
    public void Test2()
    {
      Client c1 = new Client(1);
      List<Transaction> trans1 = new List<Transaction>();
      trans1.Add(new Transaction(1, 10, new ItemGroup(new List<int> { 1, 2 })));
      trans1.Add(new Transaction(1, 25, new ItemGroup(new List<int> { 4, 6 })));
      trans1.Add(new Transaction(1, 45, new ItemGroup(new List<int> { 3 })));
      trans1.Add(new Transaction(1, 50, new ItemGroup(new List<int> { 1, 2 })));
      trans1.Add(new Transaction(1, 65, new ItemGroup(new List<int> { 3 })));
      trans1.Add(new Transaction(1, 90, new ItemGroup(new List<int> { 2, 4 })));
      trans1.Add(new Transaction(1, 95, new ItemGroup(new List<int> { 6 })));
      c1.SetTransactions(trans1, 7);
      min_gap = 5;
      max_gap = 30;
      win_size = 0;

      Sequence seq = new Sequence(new List<ItemGroup> { new ItemGroup( new List<int> { 1, 2 } ),
                                                        new ItemGroup( new List<int> { 3 } ),
                                                        new ItemGroup( new List<int> { 4 } ) });

      bool res = this.ClientSupportsSequence(c1, seq);
    }
    public void Test3()
    {
      Client c1 = new Client(1);
      List<Transaction> trans1 = new List<Transaction>();
      trans1.Add(new Transaction(1, 10, new ItemGroup(new List<int> { 1 })));
      trans1.Add(new Transaction(1, 20, new ItemGroup(new List<int> { 3 })));
      trans1.Add(new Transaction(1, 26, new ItemGroup(new List<int> { 2 })));
      c1.SetTransactions(trans1, 3);
      this.clients = new List<Client>();
      this.clients.Add(c1);
      item_map.Add(1, "11");
      item_map.Add(2, "22");
      item_map.Add(3, "33");
      min_gap = 5; //5 or 10
      max_gap = 15; //15 or 5
      win_size = 0; //0 or 6
      support = 1;

      List<Sequence> res = GetFrequentSequences();
      this.PrintFrequentSequences();
    }
    public void Test4()
    {
      LogReader log_reader = new LogReader();
      log_reader.ReadLog("C:\\Users\\arsen\\AppData\\Roaming\\Nanosoft\\nanoCAD x64 22.0\\nnmsstcs\\2022-09.log");
      log_reader.GetDataForGPS(out this.clients, out this.item_map);
      min_gap = 1;
      max_gap = 15;
      win_size = 0;
      support = 1;

      List<Sequence> res = GetFrequentSequences();
      this.PrintFrequentSequences();
      this.WriteFrequentSequencesToFile();
    }
    public void Test5()
    {
      LogReader log_reader = new LogReader();
      log_reader.ReadDir(".\\logs");
      log_reader.GetDataForGPS(out this.clients, out this.item_map);
      min_gap = 1;
      max_gap = 15;
      win_size = 0;
      support = 0.5;

      List<Sequence> res = GetFrequentSequences();
      int max_len = this.PrintFrequentSequences();
      this.WriteFrequentSequencesToFile("gsp_result.txt", max_len);
    }
  }
}

File: GSP_adapt/GSP.cpp
#include "GSP.h"
#include <chrono>

typedef chrono::high_resolution_clock Clock;

GSP::GSP()
{
}

QList<Sequence> GSP::getFrequentSequences(double _min_sup, int _min_gap, int _max_gap)
{
  if (_min_sup >= 0)
    min_support = _min_sup;
  if (_min_gap >= 0)
    min_gap = _min_gap;
  if (_max_gap >= 0)
    max_gap = _max_gap;

  prepareGSP();
  //ToDo Заменить на постепенную подгрузу сессий по страницам
  QList<Session> sessions;
  shared_ptr<DataBase> db = DataBase::instance();
  db->getAllLogs(cmds_map.size(), sessions);

  this->freq_seqs.clear();
  QList<Sequence> candidates = generateCandidates1();
  int added_seqs_num = countSupport(candidates, sessions);
  if (freq_seqs.size() > 0)
  {
    while (added_seqs_num > 0)
    {
      candidates = generateCandidates();
      qDebug() << "generateCandidates() candidates.size() = " << candidates.size();
      if (candidates.size() > 0)
        added_seqs_num = countSupport(candidates, sessions);
      else
        break;
      qDebug() << "countSupport() added_seqs_num = " << added_seqs_num;
    }
  }
  sortFrequentSequences();
  return this->freq_seqs;
}

void GSP::printFrequentSequences()
{
  /*for (int i = 0; i < freq_seqs.size(); i++)
    cout << freq_seqs[i];*/

  //get max_len
  int max_len = 0;
  QString res = "";
  for (int i = 0; i < freq_seqs.size(); i++)
  {
    res = "<";
    for (int j = 0; j < freq_seqs[i].size(); j++)
      res += cmds_map.value(freq_seqs[i][j]) + ", ";
    res.remove(res.size() - 2, 2);
    res += "> ";
    if (res.size() > max_len)
      max_len = res.size();
  }

  //print
  res = "";
  for (int i = 0; i < freq_seqs.size(); i++)
  {
    res = "<";
    for (int j = 0; j < freq_seqs[i].size(); j++)
      res += cmds_map.value(freq_seqs[i][j]) + ", ";
    res.remove(res.size() - 2, 2);
    res += ">";
    while (res.size() < max_len)
      res += " ";
    res+= "supp = " + QString::number(freq_seqs[i].support, 'f', 3);
    res+= " lift = " + (freq_seqs[i].lift < 0 ? "null ": QString::number(freq_seqs[i].lift, 'f', 3));
    qDebug() << res;
  }
}

QString GSP::getSeqStr(const Sequence &seq)
{
  QString res = "<";
  for (int j = 0; j < seq.size(); j++)
    res += cmds_map.value(seq[j]) + ", ";
  res.remove(res.size() - 2, 2);
  res += ">";
  return res;
}

void GSP::setSameCmds(bool val)
{
  same_cmds = val;
}

void GSP::test1()
{
  cur_freq_seqs.clear();
  cur_freq_seqs.push_back({1, 2, 3});
  cur_freq_seqs.push_back({1, 2, 4});
  cur_freq_seqs.push_back({1, 3, 4});
  cur_freq_seqs.push_back({1, 3, 5});
  cur_freq_seqs.push_back({2, 3, 4});
  cur_freq_seqs.push_back({2, 3, 5});

  QList<Sequence> candidates = this->generateCandidates();

  for (int i = 0; i < candidates.size(); i++)
    cout << candidates[i];
}

void GSP::test5()
{
  shared_ptr<LogReader> log_reader = LogReader::instance();

  shared_ptr<DataBase> db = DataBase::instance();
  db->setSQLiteDataBase();
  db->resetSQLiteDataBase();
  chrono::time_point<Clock> start = Clock::now();
  log_reader->readLogs(".\\logs");
  chrono::time_point<Clock> end = Clock::now();
  chrono::nanoseconds diff = chrono::duration_cast<chrono::nanoseconds>(end - start);
  qDebug() << "readLogs() time: " << diff.count() / 1000000000. << " s";

  min_gap = 0; // 0 1
  max_gap = INT_MAX; // 15 INT_MAX
  min_support = 0.5; // 0.5 0.01

  start = Clock::now();
  QList<Sequence> res = getFrequentSequences();
  end = Clock::now();
  diff = chrono::duration_cast<chrono::nanoseconds>(end - start);
  qDebug() << "getFrequentSequences() time: " << diff.count() / 1000000000. << " s";

  start = Clock::now();
  printFrequentSequences();
  end = Clock::now();
  diff = chrono::duration_cast<chrono::nanoseconds>(end - start);
  qDebug() << "printFrequentSequences() time: " << diff.count() / 1000000000. << " s";
}

void GSP::test6()
{
  shared_ptr<LogReader> log_reader = LogReader::instance();
  log_reader->readLogsWithoutTime(".\\logs_without_time");
  min_gap = 1;
  max_gap = 15;
  min_support = 0.5;

  QList<Sequence> res = getFrequentSequences();
  printFrequentSequences();
}

void GSP::prepareGSP()
{
  shared_ptr<DataBase> db = DataBase::instance();
  if (db->getCmdsMap(cmds_map) != OK)
    throw db->lastError();

  if (db->getSessionsNum(sessions_count) != OK)
    throw db->lastError();
}

QList<Sequence> GSP::generateCandidates1()
{
  QList<Sequence> candidates;
  for (int i = 0; i < this->cmds_map.size(); i++)
    candidates.push_back(Sequence(i));
  return candidates;
}

QList<Sequence> GSP::generateCandidates()
{
  QList<Sequence> candidates;
  //Join
  for (const Sequence &seq1 : cur_freq_seqs)
    for (const Sequence &seq2 : cur_freq_seqs)
    {
      //if (seq1 != seq2 && Sequence::isJoinable(seq1, seq2)) // excludes <(x),(x)>
      if (Sequence::isJoinable(seq1, seq2) && (same_cmds || seq1 != seq2))
        candidates.push_back(Sequence::join(seq1, seq2)); // <(x),(y)>
    }
  return candidates;
}

bool GSP::findCommand(int cmd, const Session &session, int min_time, int prev_cmd_id, int& time, int &id) const
{
  const QList<forward_list<pair<int,int>>> nodes_list = session.getRepresintationNodesList();
  int item_id = cmd;
  forward_list<pair<int,int>>::const_iterator it = nodes_list[item_id].cbegin();
  if (it._M_node == NULL)
    return false;
  time = -1;
  id = -1;
  while (it._M_node != NULL && it->first < min_time)
    it++;

  //Если id предыдущей команды больше чем id найденной команды, то идем дальше, т.к. не соблюден порядок, хоть и время одинаковое
  if (prev_cmd_id >= 0)
    while (it._M_node != NULL && prev_cmd_id >= it->second)
      it++;

  if (it._M_node == NULL)
    return false;
  else
  {
    time = it->first;
    id = it->second;
    return true;
  }
}

bool GSP::sessionSupportsSequence(const Session& session, const Sequence& seq)
{
  bool res = false;
  int phase = 1;
  int cur_id = 0;
  int times[seq.size()];
  int new_time;
  int new_id;
  int cmds[seq.size()];
  int ids[seq.size()];

  while (phase != 0)
  {
    if (cur_id == seq.size())
    {
      res = true;
      break;
    }
    if (phase == 1) //Forward phase
    {
      int prev_cmd_id = cur_id == 0 ? -1 : ids[cur_id - 1];
      int min_time = cur_id == 0 ? 0 : times[cur_id - 1] + min_gap;
      bool is_founded = findCommand(seq[cur_id], session, min_time, prev_cmd_id, new_time, new_id);
      if (is_founded)
      {
        times[cur_id] = new_time;
        cmds[cur_id] = seq[cur_id];
        ids[cur_id] = new_id;
      }
      else // no command in session
      {
        res = false;
        phase = 0;
        break;
      }
      if (cur_id == 0 || new_time - times[cur_id - 1] <= max_gap) //ok
      {
        cur_id++;
        phase = 1;
      }
      else // switch to backward phase
        phase = 2;
    }
    else if (phase == 2) //Backward phase
    {
      if (cur_id == 0)
        phase = 1;
      else
      {
        int prev_cmd_id = cur_id < 2 ? -1 : ids[cur_id - 2];
        int min_time = times[cur_id] - max_gap;
        bool is_founded = findCommand(seq[cur_id - 1], session, min_time, prev_cmd_id, new_time, new_id);
        if (is_founded)
        {
          times[cur_id - 1] = new_time;
          cmds[cur_id - 1] = seq[cur_id - 1];
          ids[cur_id - 1] = new_id;
        }
        else // no command in session
        {
          res = false;
          phase = 0;
          break;
        }
        if (cur_id - 1 == 0 || new_time - times[cur_id - 2] <= max_gap) //ok
        {
          if (times[cur_id] - new_time >= min_gap && new_id < ids[cur_id])
            cur_id++;
          phase = 1;
        }
        else // continue backward phase
        {
          cur_id--;
          phase = 2;
        }
      }
    }
    else //eror phase, shouldn't be here
      break;
  }

  return res;
}

Sequence GSP::findFreqSequenceByCommand(int cmd)
{
  for (int i = 0; i < freq_seqs.size(); i++)
  {
    if (freq_seqs[i] == cmd)
      return freq_seqs[i];
  }
  return Sequence();
}

double GSP::calcLift(Sequence seq)
{
  double res = seq.support;
  for (int i = 0; i < seq.size(); i++)
  {
    Sequence subseq = findFreqSequenceByCommand(seq[i]);
    if (subseq.size() < 1)
      return -1;
    res /= subseq.support;
  }
  return res;
}

int GSP::countSupport(QList<Sequence> &candidates, const QList<Session>& sessions)
{
  int res = 0;
  cur_freq_seqs.clear();
  for (Sequence& candidate : candidates)
  {
    for (const Session& session : sessions)
    {
      if (sessionSupportsSequence(session, candidate))
        candidate.support++;
    }
    candidate.support = candidate.support / sessions_count;
    //candidate.confidence = calcConfidence(candidate);
    candidate.lift = calcLift(candidate);
    if (candidate.support >= this->min_support)
    {
      freq_seqs.push_back(candidate);
      cur_freq_seqs.push_back(candidate);
      res++;
    }
  }
  return res;
}

void GSP::sortFrequentSequences()
{
  sort(freq_seqs.begin(), freq_seqs.end(), greater<>());
}
